apps-rendering/src/outline.ts

```diff
@@ -0,0 +1,57 @@
+import { BodyElement } from 'bodyElement';
+import { ElementKind } from 'bodyElementKind';
+import { indexOptional } from 'lib';
+import { Optional } from 'optional';
+
+type Fields = {
+	id: string;
+	doc: Node;
+};
+
+type Outline = Array<
+	Fields & {
+		subheadings: Array<Fields>;
+	}
+>;
+
+const fromBodyElements = (elements: BodyElement[]): Outline => {
+	return elements.reduce(
+		(outline: Outline, element: BodyElement): Outline => {
+			switch (element.kind) {
+				case ElementKind.HeadingTwo:
+					return element.id
+						.map((id) => [
+							...outline,
+							{
+								id,
+								doc: element.doc,
+								subheadings: [],
+							},
+						])
+						.withDefault(outline);
+
+				case ElementKind.HeadingThree:
+					const last = indexOptional(outline.length - 1)(outline);
+					return Optional.map2(last, element.id, (l, id) => [
+						...outline.slice(0, outline.length - 1),
+						{
+							...l,
+							subheadings: [
+								...l.subheadings,
+								{
+									id,
+									doc: element.doc,
+								},
+							],
+						},
+					]).withDefault(outline);
+				default:
+					return outline;
+			}
+		},
+		[],
+	);
+};
```

Comment:
Couple of small suggestions that may make this easier to read?

1. This function only returns one thing, so we can drop the `{}` and `return`. This removes some of the indentation.
2. Perhaps we can rename `last` and `l` to `currentH2` and `h2`, to make it a bit clearer what they're referring to?

```suggestion
const fromBodyElements = (elements: BodyElement[]): Outline =>
	elements.reduce((outline: Outline, element: BodyElement): Outline => {
		switch (element.kind) {
			case ElementKind.HeadingTwo:
				return element.id
					.map((id) => [
						...outline,
						{
							id,
							doc: element.doc,
							subheadings: [],
						},
					])
					.withDefault(outline);
			case ElementKind.HeadingThree:
				const currentH2 = indexOptional(outline.length - 1)(outline);
				return Optional.map2(currentH2, element.id, (h2, id) => [
					...outline.slice(0, outline.length - 1),
					{
						...h2,
						subheadings: [
							...h2.subheadings,
							{
								id,
								doc: element.doc,
							},
						],
					},
				]).withDefault(outline);
			default:
				return outline;
		}
	}, []);
```

Reply:
Yea certainly true. Done 👍